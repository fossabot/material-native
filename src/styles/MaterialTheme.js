'use strict';
import * as colors from './colors';
import shades from './shades';
const F = Object.freeze;

function normalizeTheme(theme) {
	return theme === 'dark' ? 'dark' : 'light';
}

function inverse(theme) {
	return theme === 'dark' ? 'light' : 'dark';
}

const themeTones = F({
	light: F({
		statusBar: colors.grey300,
		barStyle: 'dark-content',
		appBar: colors.grey100,
		background: colors.grey50,
		container: colors.white,
	}),
	dark: F({
		statusBar: colors.black,
		barStyle: 'light-content',
		appBar: colors.grey900,
		background: '#303030',
		container: colors.grey800,
	}),
});

export default class MaterialTheme {
	static defaultTheme = new MaterialTheme();

	constructor(settings) {
		this.settings = settings || {};
		// @todo recursively freeze settings
		this._applyDefaults();
		this._buildTheme();
		this._freeze();
	}

	_applyDefaults() {
		this.settings.theme = normalizeTheme(this.settings.theme);
	}

	_buildTheme() {
		const {theme} = this.settings;
		const inverseTheme = inverse(theme);

		this.themeTone = theme;
		this.inverseThemeTone = inverseTheme;

		this.palette = {
			primary: this.settings.primary || colors.indigo500,
			darkPrimary: this.settings.darkPrimary || colors.indigo700,
			accent: this.settings.accent || colors.pinkA200,
			background: themeTones[theme].background,
			container: themeTones[theme].container,
		};

		this.defaultToolbar = Object.assign({
			theme,
			background: themeTones[theme].appBar,
		}, this.settings.defaultToolbar || {});

		// @todo Automatically pick the right appBar theme based on the appBar background
		//       Many solid appBar colors should implicitly turn the theme within the appBar to dark
		const appBarTheme = normalizeTheme((this.settings.appBar && this.settings.appBar.theme) || this.defaultToolbar.theme);
		this.appBar = Object.assign({}, this.defaultToolbar, {
			theme: appBarTheme,
			background: this.palette.primary || this.defaultToolbar.background,
		}, this.settings.appBar || {});

		this.statusBar = Object.assign({
			background: this.palette.darkPrimary || themeTones[inverse(appBarTheme)].statusBar,
			translucentBackground: shades.dark.statusBar,
			barStyle: themeTones[inverse(appBarTheme)].barStyle,
		}, this.settings.statusBar || {});

		this.text = Object.assign({
			primaryColor: shades[inverseTheme].primaryText,
			secondaryColor: shades[inverseTheme].secondaryText,
			disabledColor: shades[inverseTheme].disabledText,
			hintColor: shades[inverseTheme].hintText,
			primaryOpacity: shades.opacity[inverseTheme].primaryText,
			secondaryOpacity: shades.opacity[inverseTheme].secondaryText,
			disabledOpacity: shades.opacity[inverseTheme].disabledText,
			hintOpacity: shades.opacity[inverseTheme].hintText,
		}, this.settings.text || {});

		this.icon = Object.assign({
			activeColor: shades[inverseTheme].activeIcon,
			inactiveColor: shades[inverseTheme].inactiveIcon,
			activeOpacity: shades.opacity[inverseTheme].activeIcon,
			inactiveOpacity: shades.opacity[inverseTheme].inactiveIcon,
		}, this.settings.icon || {});
	}

	_freeze() {
		Object.freeze(this); // @todo recursive
	}

	/**
	 * Create a new theme object that inherits settings from another
	 */
	extend(settings) {
		// @todo immutably merge `this.settings` with `settings` and pass that to `new MaterialTheme()`
	}
}

export function getMaterialTheme(settings) {
	return new MaterialTheme(settings);
}
