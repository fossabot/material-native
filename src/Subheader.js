'use strict';
import React, {PureComponent} from 'react';
import {StyleSheet, View, Text} from 'react-native';
import withMaterialTheme from './styles/withMaterialTheme';
import * as typo from './styles/typo';
import shades from './styles/shades';

/**
 * Section heading
 */
class Subheader extends PureComponent {
	static defaultProps = {
		inset: false,
		lines: 1,
	};

	_setRef = (ref) => this._ref = ref;
	measure(...args) { return this._ref.measure(...args); }
	measureInWindow(...args) { return this._ref.measureInWindow(...args); }
	measureLayout(...args) { return this._ref.measureLayout(...args); }

	render() {
		const {
			materialTheme,
			style,
			textStyle,
			inset,
			text,
			lines,
			secondary,
			color: colorOverride,
			dark,
			light,
			...textProps
		} = this.props;

		let color;
		if ( colorOverride ) {
			color = colorOverride;
		} else if ( dark || light ) {
			const theme = dark ? 'dark' : 'light';
			if ( secondary ) color = shades[theme].secondaryText;
			else color = shades[theme].primaryText;
		} else {
			if ( secondary ) color = materialTheme.text.secondaryColor;
			else color = materialTheme.text.primaryColor;
		}

		return (
			<View
				ref={this._setRef}
				style={[
					styles.container,
					inset && styles.inset,
					style,
					styles.containerOverrides
				]}>
				<Text
					{...textProps}
					numberOfLines={lines}
					style={[
						styles.text,
						textStyle,
						{color}
					]}>
					{text}
				</Text>
			</View>
		);
	}
}

export default withMaterialTheme(Subheader);

const styles = StyleSheet.create({
	container: {
		height: 48,
		paddingHorizontal: 16,
	},
	containerOverrides: {
		flexDirection: 'row',
		alignItems: 'center',
	},
	inset: {
		paddingRight: 16,
		paddingLeft: 72,
	},
	text: {
		...typo.fontMedium,
		fontSize: 14,
	},
});
