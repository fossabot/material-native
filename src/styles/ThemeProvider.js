'use strict';
import Observable from 'zen-observable';
import PropTypes from '../PropTypes';
import React, {PureComponent} from 'react';
import MaterialTheme from './MaterialTheme';

const defaultTheme = new MaterialTheme();

export default class ThemeProvider extends PureComponent {
	static contextTypes = {
		materialThemeObservable: PropTypes.Observable,
	};

	static childContextTypes = {
		materialThemeObservable: PropTypes.Observable.isRequired,
		memoizedMaterialStyles: PropTypes.func,
	};

	getChildContext() {
		return {
			materialThemeObservable: this.observable,
			memoizedMaterialStyles: this.memoizedMaterialStyles,
		};
	}

	componentWillMount() {
		// materialTheme
		if ( this.context.materialThemeObservable ) {
			this._parentSubscription = this.context.materialThemeObservable.subscribe({
				next: (theme) => {
					this.parentMaterialTheme = theme;
					this.handleThemeChange(this.props);
				},
				complete: () => {
					delete this._parentSubscription;
				},
			});
		} else {
			this.handleThemeChange(this.props);
		}
	}

	componentWillReceiveProps(nextProps) {
		// materialTheme
		this.handleThemeChange(nextProps);
	}

	componentWillUnmount() {
		if ( this._parentSubscription ) {
			this._parentSubscription.unsubscribe();
		}
	}

	// materialTheme
	observable = new Observable((observer) => {
		if ( this.materialTheme ) {
			observer.next(this.materialTheme);
		}

		this.observers.add(observer);

		return () => {
			this.observers.delete(observer);
		};
	});

	observers = new Set();

	materialTheme = undefined;
	parentMaterialTheme = undefined;

	// materialTheme
	handleThemeChange(props) {
		const baseTheme = props.theme || this.parentMaterialTheme || defaultTheme;

		// @todo Handle extend prop here

		if ( baseTheme !== this.materialTheme ) {
			this.materialTheme = baseTheme;
			this.observers.forEach((observer) => observer.next(this.materialTheme));
		}
	}

	// styleCache
	materialStyleCache = new Map();

	memoizedMaterialStyles = (materialTheme, mapThemeToStyles) => {
		let cache = this.materialStyleCache.get(mapThemeToStyles);
		if ( cache && cache.materialTheme === materialTheme ) {
			return cache.styles;
		}

		// @note If possible in the future, styles in `cache` can be unloaded at this point

		const cacheStyles = mapThemeToStyles(materialTheme);

		for ( const key in cacheStyles ) {
			Object.freeze(cacheStyles[key]);
		}
		Object.freeze(cacheStyles);

		cache = {
			materialTheme,
			styles: cacheStyles,
		};

		this.materialStyleCache.set(mapThemeToStyles, cache);

		return cache.styles;
	};

	// boilerplate
	render() {
		return React.Children.only(this.props.children);
	}
}
