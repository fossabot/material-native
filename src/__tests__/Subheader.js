'use strict';
import React from 'react';
import MaterialTheme from '../styles/MaterialTheme';
import ThemeProvider from '../styles/ThemeProvider';
import Subheader from '../Subheader';

import ReactTestRenderer from 'react-test-renderer';
import ToJSX from 'react-element-to-jsx-string';

const theme = new MaterialTheme({});

/* eslint-disable react/jsx-key, react-native/no-inline-styles */
const snapshotTests = [
	<Subheader text='Subhead' />,
	<Subheader text='Subhead' onLayout={() => {}} />,
	<Subheader text='Subhead' inset />,
	<Subheader text='Subhead' lines={2} />,
	<Subheader text='Subhead' dark />,
	<Subheader text='Subhead' light />,
	<Subheader text='Subhead' secondary />,
	<Subheader text='Subhead' secondary dark />,
	<Subheader text='Subhead' secondary light />,
	<Subheader text='Subhead' color='#123456' />,
	<Subheader text='Subhead' textStyle={{width: 100}} />
];
/* eslint-enable react/jsx-key, react-native/no-inline-styles */

for ( const root of snapshotTests ) {
	test(ToJSX(root, {maxInlineAttributesLineLength: Infinity}), () => {
		const host = ReactTestRenderer.create(
			<ThemeProvider theme={theme}>
				{root}
			</ThemeProvider>
		);
		expect(host.toJSON()).toMatchSnapshot();
	});
}

test("<Subheader text='Subhead' /> to have measure methods", () => {
	let instance;
	ReactTestRenderer.create(
		<ThemeProvider theme={theme}>
			<Subheader text='Subhead' ref={(ref) => instance = ref} />
		</ThemeProvider>
	);

	expect(instance).toBeDefined();
	expect(instance.measure).toBeInstanceOf(Function);
	expect(instance.measureInWindow).toBeInstanceOf(Function);
	expect(instance.measureLayout).toBeInstanceOf(Function);
});
