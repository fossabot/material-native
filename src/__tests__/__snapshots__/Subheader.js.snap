// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<withMaterialTheme(Subheader)
  text="Subhead"
  textStyle={{
    width: 100
  }}
 /> 1`] = `
<View
  style={
    Array [
      Object {
        "height": 48,
        "paddingHorizontal": 16,
      },
      false,
      undefined,
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    numberOfLines={1}
    style={
      Array [
        Object {
          "fontFamily": "System",
          "fontSize": 14,
          "fontWeight": "500",
        },
        Object {
          "width": 100,
        },
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
        },
      ]
    }
  >
    Subhead
  </Text>
</View>
`;

exports[`<withMaterialTheme(Subheader) color="#123456" text="Subhead" /> 1`] = `
<View
  style={
    Array [
      Object {
        "height": 48,
        "paddingHorizontal": 16,
      },
      false,
      undefined,
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    numberOfLines={1}
    style={
      Array [
        Object {
          "fontFamily": "System",
          "fontSize": 14,
          "fontWeight": "500",
        },
        undefined,
        Object {
          "color": "#123456",
        },
      ]
    }
  >
    Subhead
  </Text>
</View>
`;

exports[`<withMaterialTheme(Subheader) dark secondary text="Subhead" /> 1`] = `
<View
  style={
    Array [
      Object {
        "height": 48,
        "paddingHorizontal": 16,
      },
      false,
      undefined,
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    numberOfLines={1}
    style={
      Array [
        Object {
          "fontFamily": "System",
          "fontSize": 14,
          "fontWeight": "500",
        },
        undefined,
        Object {
          "color": "rgba(0, 0, 0, 0.54)",
        },
      ]
    }
  >
    Subhead
  </Text>
</View>
`;

exports[`<withMaterialTheme(Subheader) dark text="Subhead" /> 1`] = `
<View
  style={
    Array [
      Object {
        "height": 48,
        "paddingHorizontal": 16,
      },
      false,
      undefined,
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    numberOfLines={1}
    style={
      Array [
        Object {
          "fontFamily": "System",
          "fontSize": 14,
          "fontWeight": "500",
        },
        undefined,
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
        },
      ]
    }
  >
    Subhead
  </Text>
</View>
`;

exports[`<withMaterialTheme(Subheader) inset text="Subhead" /> 1`] = `
<View
  style={
    Array [
      Object {
        "height": 48,
        "paddingHorizontal": 16,
      },
      Object {
        "paddingLeft": 72,
        "paddingRight": 16,
      },
      undefined,
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    numberOfLines={1}
    style={
      Array [
        Object {
          "fontFamily": "System",
          "fontSize": 14,
          "fontWeight": "500",
        },
        undefined,
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
        },
      ]
    }
  >
    Subhead
  </Text>
</View>
`;

exports[`<withMaterialTheme(Subheader) light secondary text="Subhead" /> 1`] = `
<View
  style={
    Array [
      Object {
        "height": 48,
        "paddingHorizontal": 16,
      },
      false,
      undefined,
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    numberOfLines={1}
    style={
      Array [
        Object {
          "fontFamily": "System",
          "fontSize": 14,
          "fontWeight": "500",
        },
        undefined,
        Object {
          "color": "rgba(255, 255, 255, 0.7)",
        },
      ]
    }
  >
    Subhead
  </Text>
</View>
`;

exports[`<withMaterialTheme(Subheader) light text="Subhead" /> 1`] = `
<View
  style={
    Array [
      Object {
        "height": 48,
        "paddingHorizontal": 16,
      },
      false,
      undefined,
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    numberOfLines={1}
    style={
      Array [
        Object {
          "fontFamily": "System",
          "fontSize": 14,
          "fontWeight": "500",
        },
        undefined,
        Object {
          "color": "rgba(255, 255, 255, 1)",
        },
      ]
    }
  >
    Subhead
  </Text>
</View>
`;

exports[`<withMaterialTheme(Subheader) lines={2} text="Subhead" /> 1`] = `
<View
  style={
    Array [
      Object {
        "height": 48,
        "paddingHorizontal": 16,
      },
      false,
      undefined,
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    numberOfLines={2}
    style={
      Array [
        Object {
          "fontFamily": "System",
          "fontSize": 14,
          "fontWeight": "500",
        },
        undefined,
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
        },
      ]
    }
  >
    Subhead
  </Text>
</View>
`;

exports[`<withMaterialTheme(Subheader) onLayout={function noRefCheck() {}} text="Subhead" /> 1`] = `
<View
  style={
    Array [
      Object {
        "height": 48,
        "paddingHorizontal": 16,
      },
      false,
      undefined,
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    numberOfLines={1}
    onLayout={[Function]}
    style={
      Array [
        Object {
          "fontFamily": "System",
          "fontSize": 14,
          "fontWeight": "500",
        },
        undefined,
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
        },
      ]
    }
  >
    Subhead
  </Text>
</View>
`;

exports[`<withMaterialTheme(Subheader) secondary text="Subhead" /> 1`] = `
<View
  style={
    Array [
      Object {
        "height": 48,
        "paddingHorizontal": 16,
      },
      false,
      undefined,
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    numberOfLines={1}
    style={
      Array [
        Object {
          "fontFamily": "System",
          "fontSize": 14,
          "fontWeight": "500",
        },
        undefined,
        Object {
          "color": "rgba(0, 0, 0, 0.54)",
        },
      ]
    }
  >
    Subhead
  </Text>
</View>
`;

exports[`<withMaterialTheme(Subheader) text="Subhead" /> 1`] = `
<View
  style={
    Array [
      Object {
        "height": 48,
        "paddingHorizontal": 16,
      },
      false,
      undefined,
      Object {
        "alignItems": "center",
        "flexDirection": "row",
      },
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    disabled={false}
    ellipsizeMode="tail"
    numberOfLines={1}
    style={
      Array [
        Object {
          "fontFamily": "System",
          "fontSize": 14,
          "fontWeight": "500",
        },
        undefined,
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
        },
      ]
    }
  >
    Subhead
  </Text>
</View>
`;
